### PR TITLE
fix(api): optimise archived list query. Fixes #13295

### LINKED ISSUE #13295: 3.5: Improve Archived API/DB read performance
---
_edited by agilgur5_: 
- Please help test #13566 by running [the test image below](https://github.com/argoproj/argo-workflows/issues/13295#issuecomment-2339407775)

---

### Pre-requisites

- [X] I have double-checked my configuration
- [X] I have tested with the `:latest` image tag (i.e. `quay.io/argoproj/workflow-controller:latest`) and can confirm the issue still exists on `:latest`. If not, I have explained why, **in detail**, in my description below.
- [X] I have searched existing issues and could not find a match for this bug
- [X] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/main/docs/CONTRIBUTING.md))

### What happened/what did you expect to happen?

Currently , when large number of workflows are running , Workflows view of Argo workflows UI is taking up to 12 seconds to load. I don't find any way to further optimise it. CPU and Memory uses of server pod are normal. I am using Postgres to archive the workflows and cpu and memory usage of database as well are  normal. i don't see any spikes anywhere. So looks like its not resource contention but something else,   What can be done to improve the load time .

### Version

3.5.0, 3.5.6, 3.5.8,

### Paste a small workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
Its UI load time (workflow view)
```

### Logs from the workflow controller

```text
Its UI load time (workflow view)
```

### Logs from in your workflow's wait container

```text
Its UI load time (workflow view)
```

### PR DESCRIPTION
**Pull request by bom-d-van**
Fixes #13295

### Motivation

The current implementation appears to be causing postgres always unmarshalling workflow json payload for all the records in the table. by adopting a subquery approach, we are able to optimise the query from a runtime of 11495.734 ms to 44.713 ms. the data size is about 417481 argo_archived_workflows and 1794624 argo_archived_workflows_labels.

this new change is backward compatible and it is tested on our production env (using postgresql).

### Modifications

rewriting the generated query to use a subquery without unmarshalling json payload first (effectively reducing workload required from postgres, although arguably postgres should be smart enough to only perform the operation after querying ;)):

previous:

```sql
SELECT name,
       namespace,
       UID,
       phase,
       startedat,
       finishedat,
       coalesce((workflow::JSON)->'metadata'->>'labels', '{}') AS labels,
       coalesce((workflow::JSON)->'metadata'->>'annotations', '{}') AS annotations,
       coalesce((workflow::JSON)->'status'->>'progress', '') AS progress,
       coalesce((workflow::JSON)->'metadata'->>'creationTimestamp', '') AS creationtimestamp,
       (workflow::JSON)->'spec'->>'suspend' AS suspend,
       coalesce((workflow::JSON)->'status'->>'message', '') AS message,
       coalesce((workflow::JSON)->'status'->>'estimatedDuration', '0') AS estimatedduration,
       coalesce((workflow::JSON)->'status'->>'resourcesDuration', '{}') AS resourcesduration
FROM "argo_archived_workflows"
WHERE (("clustername" = 'default'
        AND "instanceid" = '')
       AND "namespace" = 'argo-map'
       AND EXISTS
         (SELECT 1
          FROM argo_archived_workflows_labels
          WHERE clustername = argo_archived_workflows.clustername
            AND UID = argo_archived_workflows.uid
            AND name = 'workflows.argoproj.io/phase'
            AND value = 'Succeeded')
       AND EXISTS
         (SELECT 1
          FROM argo_archived_workflows_labels
          WHERE clustername = argo_archived_workflows.clustername
            AND UID = argo_archived_workflows.uid
            AND name = 'workflows.argoproj.io/workflow-template'
            AND value = 'mapping1-pipeline-template-with-nfs'))
ORDER BY "startedat" DESC
LIMIT 1;
```

now:

```sql
SELECT name,
       namespace,
       UID,
       phase,
       startedat,
       finishedat,
       coalesce((workflow::JSON)->'metadata'->>'labels', '{}') AS labels,
       coalesce((workflow::JSON)->'metadata'->>'annotations', '{}') AS annotations,
       coalesce((workflow::JSON)->'status'->>'progress', '') AS progress,
       coalesce((workflow::JSON)->'metadata'->>'creationTimestamp', '') AS creationtimestamp,
       (workflow::JSON)->'spec'->>'suspend' AS suspend,
       coalesce((workflow::JSON)->'status'->>'message', '') AS message,
       coalesce((workflow::JSON)->'status'->>'estimatedDuration', '0') AS estimatedduration,
       coalesce((workflow::JSON)->'status'->>'resourcesDuration', '{}') AS resourcesduration
FROM "argo_archived_workflows"
WHERE "clustername" = 'default'
  AND UID IN
    (SELECT UID
     FROM "argo_archived_workflows"
     WHERE (("clustername" = 'default'
             AND "instanceid" = '')
            AND "namespace" = 'argo-map'
            AND EXISTS
              (SELECT 1
               FROM argo_archived_workflows_labels
               WHERE clustername = argo_archived_workflows.clustername
                 AND UID = argo_archived_workflows.uid
                 AND name = 'workflows.argoproj.io/phase'
                 AND value = 'Succeeded')
            AND EXISTS
              (SELECT 1
               FROM argo_archived_workflows_labels
               WHERE clustername = argo_archived_workflows.clustername
                 AND UID = argo_archived_workflows.uid
                 AND name = 'workflows.argoproj.io/workflow-template'
                 AND value = 'mapping1-pipeline-template-with-nfs'))
     ORDER BY "startedat" DESC
     LIMIT 1);
```

### Verification

the query runtime on our production env is reduced from 11495.734 ms to 44.713 ms. the data size is about 417481 argo_archived_workflows and 1794624 argo_archived_workflows_labels.

Screenshot 2024-08-29 at 18 51 43

Screenshot 2024-08-29 at 18 50 05

### CODE REVIEW

**Comment by bom-d-van** [comment]
hi @agilgur5  thank for the feedback. I have pushed another fix for mysql. it requires a some extra handling.

there is till one more test failure, but it seems not related to my changes. do you have any suggestions? 
or is it safe to leave it as it is.

**Comment by bom-d-van** [comment]
hmm, it seems that retry helps. :P

**Comment by agilgur5** [comment]
> there is a test failure about (`artgc-artifact-not-written-failed-zn9dv`), but it seems not related to my changes. do you have any suggestions?



> hmm, it seems that retry helps. :P

Sounds like a test flake then. Although it looks like it [was with `TestStoppedWorkflow`](https://github.com/argoproj/argo-workflows/actions/runs/10758518983/job/29834036494), and not the test you mentioned above (which was added recently):

```
artifacts_test.go:218: timeout after 1m30s waiting for condition
    artifacts_test.go:241: 
        	Error Trace:	/home/runner/work/argo-workflows/argo-workflows/test/e2e/artifacts_test.go:241
        	            				/home/runner/work/argo-workflows/argo-workflows/test/e2e/fixtures/then.go:251
        	            				/home/runner/work/argo-workflows/argo-workflows/test/e2e/artifacts_test.go:240
        	Error:      	Received unexpected error:
        	            	The specified key does not exist.
        	Test:       	TestArtifactsSuite/TestStoppedWorkflow
=== FAIL: ArtifactsSuite/TestStoppedWorkflow
```

File: persist/sqldb/workflow_archive.go [hunk 1]
```diff
@@ -163,16 +163,31 @@ func (r *workflowArchive) ListWorkflows(options sutils.ListOptions) (wfv1.Workfl
...
+	//
+	//     Error 1235 (42000): This version of MySQL doesn't yet support 'LIMIT \u0026 IN/ALL/ANY/SOME subquery'
+	//
+	// more context:
+	// * https://dev.mysql.com/doc/refman/8.0/en/subquery-errors.html
+	// * https://dev.to/gkoniaris/limit-mysql-subquery-results-inside-a-where-in-clause-using-laravel-s-eloquent-orm-26en
+	if r.dbType == MySQL {
+		subSelector = r.session.SQL().Select(db.Raw("*")).From(subSelector).As("x")
+	}
+
```
Comment (agilgur5) [inline persist/sqldb/workflow_archive.go:186]: Might be good to add a comment explaining this optimization in-line

**Review by agilgur5 — COMMENTED** [review]
This mostly looks good to me but would like some folks to double-check cc @jiachengxu @terrytangyuan @Joibel 

Also might be worthwhile to push up an image for users on the issue to test and confirm results

**Comment by agilgur5** [comment]
> the query runtime on our production env is reduced from 11495.734 ms to 44.713 ms. the data size is about 417481 argo_archived_workflows and 1794624 argo_archived_workflows_labels.

While this requires a good bit of rows to test, I'm also thinking that we could potentially write a simple test that checks if the query returns under a certain amount of time on a test seed. For example, under 100ms.
That way it could act as a regression test for anything severely off, but not a full performance test to catch everything.

**Comment by ryancurrah** [comment]
I can test it out. We have a large non-production deployment we can test it in.

**Review by jiachengxu — APPROVED** [review]

**Comment by bom-d-van** [comment]
> While this requires a good bit of rows to test, I'm also thinking that we could potentially write a simplified test that checks if the query returns under a certain amount of time on a test seed. For example, under 100ms.

@agilgur5  yeah, it sounds quite reasonable and would be helpful in the long run. but unfortunately, I didn't spot a simple way for me to add a new regression test fast enough and my recent work schedule is quite tight.

could we add a new issue and linking it to this pr for now? I might be able to make the contribution a few weeks later.

File: persist/sqldb/workflow_archive.go (hunk 1 above)
Comment (bom-d-van) [reply persist/sqldb/workflow_archive.go:186]: yep, I have moved the comments down below the if statement.

File: persist/sqldb/workflow_archive.go [hunk 2]
```diff
@@ -163,16 +163,37 @@ func (r *workflowArchive) ListWorkflows(options sutils.ListOptions) (wfv1.Workfl
...
+		// * https://dev.to/gkoniaris/limit-mysql-subquery-results-inside-a-where-in-clause-using-laravel-s-eloquent-orm-26en
+		subSelector = r.session.SQL().Select(db.Raw("*")).From(subSelector).As("x")
+	}
+
+	// why using a subquery: the json unmarshaling process of coalesce
+	// (workflow) is always triggered for every records in the filtering query
+	// in postgres (and maybe mysql as well?). by first filtering out uid first
+	// without trying to unmarshal the workflow, it kinda teach the database to
+	// delay the json parsing operation and thus speed up the query (could be
+	// up to 257 times faster for some deployments).
```
Comment (agilgur5) [inline persist/sqldb/workflow_archive.go:192]:
```suggestion
	// why a subquery? the json unmarshal triggers for every row in the filter query.
	// by filtering on uid first, we delay json parsing until a single row, speeding up the query
	// (e.g. up to 257 times faster for some deployments).
```

can simplify this a little

**Comment by agilgur5** [comment]
> could we add a new issue and linking it to this pr for now? I might be able to make the contribution a few weeks later.

Yea that'd be fine so long as we get some more manual tests of this. I'm loading up my dev env right now to push up an image

**Comment by ryancurrah** [comment]
Provide me an image to use and I will test it.

**Comment by agilgur5** [comment]
Pushed up [`agilgur5/argo-cli:fix-archive-list-query-optimization`](https://hub.docker.com/layers/agilgur5/argo-cli/fix-archive-list-query-optimization/images/sha256-d3a31a31e051a499e7b49e93a7c942414cf83eb56d6e46cc012b23758e1b36a7?context=repo) for testing

**Review by terrytangyuan — APPROVED** [review]
The change looks good.

File: persist/sqldb/workflow_archive.go [hunk 3]
```diff
@@ -163,16 +163,34 @@ func (r *workflowArchive) ListWorkflows(options sutils.ListOptions) (wfv1.Workfl
...
+		subSelector = r.session.SQL().Select(db.Raw("*")).From(subSelector).As("x")
+	}
+
+	// why a subquery? the json unmarshal triggers for every row in the filter query.
+	// by filtering on uid first, we delay json parsing until a single row, speeding up the query
+	// (e.g. up to 257 times faster for some deployments).
+	selector := r.session.SQL().Select(selectQuery).From(archiveTableName).Where(
+		r.clusterManagedNamespaceAndInstanceID().And(db.Cond{"uid IN": subSelector}),
+	)
+
```
Comment (terrytangyuan) [inline persist/sqldb/workflow_archive.go:193]: Perhaps add a link to this PR so that we can have more context

Comment (Danny5487401) [reply persist/sqldb/workflow_archive.go:193]: @terrytangyuan https://github.com/argoproj/argo-workflows/issues/13563#issuecomment-2339660938

File: persist/sqldb/workflow_archive.go [hunk 4]
```diff
@@ -163,16 +163,39 @@ func (r *workflowArchive) ListWorkflows(options sutils.ListOptions) (wfv1.Workfl
...
+		// * https://dev.to/gkoniaris/limit-mysql-subquery-results-inside-a-where-in-clause-using-laravel-s-eloquent-orm-26en
+		subSelector = r.session.SQL().Select(db.Raw("*")).From(subSelector).As("x")
+	}
+
+	// why using a subquery: the json unmarshaling process of coalesce
+	// (workflow) is always triggered for every records in the filtering query
+	// in postgres (and maybe mysql as well). by first filtering out uid first
+	// without trying to unmarshal the workflow, it kinda teach the database to
+	// delay the json parsing operation and thus speed up the query (could be
+	// up to 257 times faster for some deployments).
```
Comment (agilgur5) [inline persist/sqldb/workflow_archive.go:192]: You undid my [simplification](https://github.com/argoproj/argo-workflows/pull/13566#discussion_r1751082508) here

**Review by agilgur5 — APPROVED** [review]
Thanks for the testing @ryancurrah !

With that, some of the individual query tests folks made across both MySQL and Postgres, and 2 other approvals, I think this is good to go 🚀 

Note that this query seems to be further optimizable per https://github.com/argoproj/argo-workflows/issues/13295#issuecomment-2340964248, so we may want to open a new issue to add further optimizations, but this at least fixes the performance issues and gets us to a much more reasonable baseline

**Comment by ryancurrah** [comment]
> Thanks for the testing @ryancurrah !
> 
> With that, some of the individual query tests folks made, and 2 other approvals, I think this is good to go 🚀
> 
> Note that this query seems to be further optimizable per [#13295 (comment)](https://github.com/argoproj/argo-workflows/issues/13295#issuecomment-2340964248), so we may want to open a new issue to add further optimizations, but this at least fixes the performance issues and gets us to a much more reasonable baseline

If you need someone to test any more query optimizations feel free to ping me.

**Comment by agilgur5** [comment]
Added a follow-up issue for further optimizations in https://github.com/argoproj/argo-workflows/issues/13601

**Pull request merged**